audioswitch/src/main/java/com/twilio/audioswitch/bluetooth/BluetoothHeadsetManager.kt

```diff
@@ -48,21 +87,107 @@ internal class BluetoothHeadsetManager(
+                        when (state) {
+                            AudioManager.SCO_AUDIO_STATE_CONNECTED -> {
+                                logger.d(TAG, "Bluetooth SCO Audio connected")
+                                headsetState = AudioActivated
+                                headsetListener?.onBluetoothHeadsetStateChanged()
+                                enableBluetoothScoJob.cancelBluetoothScoJob()
+                            }
+                            AudioManager.SCO_AUDIO_STATE_DISCONNECTED -> {
+                                logger.d(TAG, "Bluetooth SCO Audio disconnected")
+                                /*
+                                 * This block is needed to restart bluetooth SCO in the event that
+                                 * the active bluetooth headset has changed.
+                                 */
+                                if (hasActiveHeadsetChanged()) {
+                                    enableBluetoothScoJob.executeBluetoothScoJob()
+                                }
+
+                                headsetListener?.onBluetoothHeadsetStateChanged()
+                                disableBluetoothScoJob.cancelBluetoothScoJob()
+                            }
+                            AudioManager.SCO_AUDIO_STATE_ERROR -> {
+                                logger.e(TAG, "Error retrieving Bluetooth SCO Audio state")
+                            }
+                        }
+                    }
+                }
+                else -> {}
+            }
+        }
+    }
+
+    fun start(headsetListener: BluetoothHeadsetConnectionListener) {
+        this.headsetListener = headsetListener
+        enableBluetoothScoJob.deviceListener = headsetListener
+
+        bluetoothAdapter.getProfileProxy(
+                context,
+                this,
+                BluetoothProfile.HEADSET)
+
+        context.registerReceiver(
+                this, IntentFilter(BluetoothDevice.ACTION_ACL_CONNECTED))
+        context.registerReceiver(
+                this, IntentFilter(BluetoothDevice.ACTION_ACL_DISCONNECTED))
+        context.registerReceiver(
+                this,
+                IntentFilter(AudioManager.ACTION_SCO_AUDIO_STATE_UPDATED))
+    }
+
     fun stop() {
         headsetListener = null
         bluetoothAdapter.closeProfileProxy(BluetoothProfile.HEADSET, headsetProxy)
+        enableBluetoothScoJob.deviceListener = null
+        context.unregisterReceiver(this)
     }
 
-    fun updateState(headsetEvent: HeadsetEvent) {
-        when (headsetEvent) {
-            Connect -> connect()
-            Disconnect -> disconnect()
-            is StartAudioActivation -> startAudioActivation(headsetEvent.error)
-            is StartAudioDeactivation -> startAudioDeactivation(headsetEvent.error)
-            HeadsetEvent.AudioActivated -> audioActivated()
-        }
+    fun activate() {
```

Comment:
Is it possible for activate to be called while in any other state? Do we need to add any illegal state exceptions here?

Reply:
At the very least it might be worth adding logging for cases where this is called in other states.